### PR TITLE
add scheduler to all components and schedule search flush after period of times

### PR DESCRIPTION
**Pull request by ekneg54**
### ToDo

- [x] add scheduler to all components to register periodic tasks and run pending tasks in `process_pipeline`
- [x] refactor output error handling in pipeline to reuse it for scheduled tasks
- [x] refactor input error handling to be concise with output error handling
- [x] refactor processor error handling to be concise with output error handling
- [x] write tests for general scheduling feature in components
- [x] write tests for registering flush timeout task in elasicsearch and opensearch output setup
- [x] add documentation for scheduling tasks for components
- [x] update changelog
- [x] reset message_backlog after it was written
- [x] rename warnings to warnings
- [x] add tests for metric counting in exceptions
- [x] revise config for quickstart to ensure outputs exists
- [x] revise dry_runner to delete redundant code for error printing

### CODE REVIEW

**Comment by ekneg54** [comment]
@ppcad I do not need a detailed code review, but could you please test this to ensure that it does not break on your site?

File: tests/unit/processor/normalizer/test_normalizer.py [hunk 1]
```diff
@@ -72,9 +73,9 @@ def test_process_normalized_field_already_exists_with_different_content(self):
         }
         with pytest.raises(
             ProcessingWarning,
```
Comment (dtrai2) [inline tests/unit/processor/normalizer/test_normalizer.py:75]: Can you specify this to `DuplicationError`?

File: tests/unit/processor/normalizer/test_normalizer.py [hunk 2]
```diff
@@ -645,9 +646,9 @@ def test_normalization_from_grok_onto_existing(self):
 
         with pytest.raises(
             ProcessingWarning,
```
Comment (dtrai2) [inline tests/unit/processor/normalizer/test_normalizer.py:648]: Can you specify this to `DuplicationError`?

File: tests/unit/framework/test_pipeline.py [hunk 3]
```diff
@@ -216,54 +232,37 @@ def test_enable_iteration_sets_iterate_to_true_stop_to_false(self, _):
...
-            "dummy": original_create({"dummy": {"type": "dummy_output"}}, mock.MagicMock())
-        }
-        self.pipeline.process_pipeline()
-        assert len(self.pipeline._output["dummy"].events) == 0
-        mock_error.assert_called()
-        assert re.search(
-            "A critical error occurred for input .*: An error message", mock_error.call_args[0][0]
-        )
-        assert len(self.pipeline._output["dummy"].failed_events) == 1
-
```
Comment (dtrai2) [inline tests/unit/framework/test_pipeline.py:236]: Why do you delete this test completely while you rewrote the test for the output connector? Can't you rewrite this one as well?

**Review by dtrai2 — CHANGES_REQUESTED** [review]
I have left you *some* comments here :wink: Mostly just small stuff though.

File: logprep/processor/base/exceptions.py [hunk 4]
```diff
@@ -45,35 +50,49 @@ def __init__(self, processor_type=None):  # pragma: no cover
...
 
-class ProcessingWarning(ProcessingError):
-    """An minor error occurred - log the error but continue processing the event."""
+class ProcessingCriticalError(ProcessingError):
+    """A critical error occured - stop processing of this event"""
 
-    def __init__(self, message: str):
-        super().__init__("ProcessingWarning", message)
+    def __init__(self, processor: "Processor", message: str, event: dict):
+        self.event = deepcopy(event)
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/processor/base/exceptions.py]: removed

File: logprep/processor/base/exceptions.py [hunk 5]
```diff
@@ -45,35 +50,49 @@ def __init__(self, processor_type=None):  # pragma: no cover
...
-class ProcessingWarningCollection(ProcessingError):
-    """A collection of ProcessingWarnings."""
+class ProcessingWarning(ProcessingError):
+    """An minor error occurred - log the error but continue processing the event."""
 
-    def __init__(self, name: str, message: str, processing_warnings):
-        super().__init__(name, message)
-        self.processing_warnings = processing_warnings
+    def __init__(self, processor: "Processor", message: str, rule: "Rule", event: dict):
+        self.event = event
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/processor/base/exceptions.py]: removed

File: tests/acceptance/test_config_refresh.py [hunk 6]
```diff
@@ -42,5 +40,5 @@ def test_no_config_refresh_after_5_seconds(tmp_path):
     wait_for_output(
         proc,
         "Configuration version didn't change. Continue running with current version.",
-        test_timeout=5,
+        test_timeout=7,
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply tests/acceptance/test_config_refresh.py:43]: yes... tests are failing sometimes on my machine

File: tests/acceptance/test_wineventlog_normalization.py [hunk 7]
```diff
@@ -29,23 +34,21 @@ def create_config():
     return get_default_logprep_config(pipeline, with_hmac=False)
 
 
-def test_events_normalized_correctly(tmp_path, config):
-    expected_output = "normalized_win_event_log.jsonl"
-    expected_output_path = path.join("tests/testdata/acceptance/expected_result", expected_output)
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply tests/acceptance/test_wineventlog_normalization.py:34]: sure -> done

File: tests/unit/component/base.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+from typing import Iterable
+from unittest import mock
+
+from logprep.abc.component import Component
+from logprep.abc.connector import Connector
+from logprep.factory import Factory
+from logprep.util.helper import camel_to_snake
+
+
+class BaseCompontentTestCase(ABC):
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply tests/unit/component/base.py]: done

File: tests/unit/framework/test_pipeline.py (hunk 3 above)
Comment (ekneg54) [reply tests/unit/framework/test_pipeline.py:236]: this was tests twice so it is redundant. look for `test_critical_input` and the first tests is the same as this

File: tests/unit/connector/test_elasticsearch_output.py [hunk 9]
```diff
@@ -228,6 +229,12 @@ def test_handle_serialization_error_raises_fatal_output_error(self):
 
     def test_setup_raises_fatal_output_error_if_unauthenticated(self):
         self.object._search_context.info = mock.MagicMock()
-        self.object._search_context.info.side_effect = elasticsearch.AuthenticationException
+        self.object._search_context.info.side_effect = elasticsearch.ElasticsearchException
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply tests/unit/connector/test_elasticsearch_output.py:232]: test renamed

File: tests/unit/exceptions/processor/test_processing_warning.py
[the remark this answers is not in the document]
Comment (ekneg54) [reply tests/unit/exceptions/processor/test_processing_warning.py:1]: they are tested in test_pipeline.py. So I sugest to do this work later. This PR is even big enough :-D

File: tests/unit/util/test_rule_dry_runner.py [hunk 10]
```diff
@@ -174,27 +175,3 @@ def test_dry_run_prints_predetection(self, tmp_path, capsys):
         assert "------ TRANSFORMED EVENTS: 1/1 ------" in captured.out
         assert "------ ALL PSEUDONYMS ------" in captured.out
         assert "------ ALL PRE-DETECTIONS ------" in captured.out
-
-    @mock.patch("logprep.processor.labeler.processor.Labeler.process", side_effect=BaseException)
-    def test_dry_run_prints_errors(self, _, tmp_path, capsys):
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply tests/unit/util/test_rule_dry_runner.py:179]: this is broken in main branch. We should not fix it here.
I created a bug ticket: #350

File: logprep/connector/elasticsearch/output.py [hunk 11]
```diff
@@ -194,25 +196,28 @@ def _write_to_search_context(self, document):
...
+                max_retries=self._config.max_retries,
+                chunk_size=self._config.message_backlog_size,
+            )
+        except elasticsearch.SerializationError as error:
+            self._handle_serialization_error(error)
+        except elasticsearch.ConnectionError as error:
+            self._handle_connection_error(error)
+        except helpers.BulkIndexError as error:
+            self._handle_bulk_index_error(error)
+        self._processed_cnt = 0
```
[the remark this answers is not in the document]
Comment (ekneg54) [reply logprep/connector/elasticsearch/output.py]: I changed the connector implementation and did some refactoring. I tested it with my local environment and it did what it should to.

File: logprep/processor/concatenator/processor.py [hunk 12]
```diff
@@ -20,7 +20,7 @@
             - tests/testdata/rules/generic/
 """
 from logprep.abc.processor import Processor
-from logprep.processor.base.exceptions import DuplicationError
+from logprep.processor.base.exceptions import FieldExsistsWarning
 from logprep.processor.concatenator.rule import ConcatenatorRule
 from logprep.util.helper import add_field_to, get_dotted_field_value
```
Comment (dtrai2) [inline logprep/processor/concatenator/processor.py:25]: Seems like you are sorting imports now, but are not yet automatically removing unused imports :wink: 
```suggestion
from logprep.processor.concatenator.rule import ConcatenatorRule
from logprep.util.helper import get_dotted_field_value
```

File: logprep/processor/datetime_extractor/processor.py [hunk 13]
```diff
@@ -26,7 +26,7 @@
 from dateutil.tz import tzlocal
 
 from logprep.abc.processor import Processor
-from logprep.processor.base.exceptions import DuplicationError
+from logprep.processor.base.exceptions import FieldExsistsWarning
 from logprep.processor.datetime_extractor.rule import DatetimeExtractorRule
 from logprep.util.helper import add_field_to, get_dotted_field_value
```
Comment (dtrai2) [inline logprep/processor/datetime_extractor/processor.py:31]:
```suggestion
from logprep.processor.datetime_extractor.rule import DatetimeExtractorRule
from logprep.util.helper import get_dotted_field_value
```

File: logprep/processor/key_checker/processor.py [hunk 14]
```diff
@@ -20,11 +20,12 @@
 """
 
 from typing import Iterable
+
 from logprep.abc.processor import Processor
+from logprep.processor.base.exceptions import FieldExsistsWarning
 from logprep.processor.base.rule import Rule
 from logprep.processor.key_checker.rule import KeyCheckerRule
 from logprep.util.helper import add_field_to, get_dotted_field_value
```
Comment (dtrai2) [inline logprep/processor/key_checker/processor.py:28]:
```suggestion
from logprep.processor.base.rule import Rule
from logprep.processor.key_checker.rule import KeyCheckerRule
from logprep.util.helper import get_dotted_field_value
```

File: tests/unit/connector/test_opensearch_output.py [hunk 15]
```diff
@@ -249,8 +210,28 @@ def test_handle_serialization_error_raises_fatal_output_error(self):
         with pytest.raises(FatalOutputError):
             self.object._handle_serialization_error(mock.MagicMock())
 
-    def test_setup_raises_fatal_output_error_if_unauthenticated(self):
+    def test_setup_raises_fatal_output_error_if_elastic_error_is_raised(self):
```
Comment (dtrai2) [inline tests/unit/connector/test_opensearch_output.py:213]:
```suggestion
    def test_setup_raises_fatal_output_error_if_opensearch_error_is_raised(self):
```

**Review by dtrai2 — CHANGES_REQUESTED** [review]
Thanks for considering all my suggestions, I have here the last few. After that I think we should be done here.

File: tests/unit/util/test_rule_dry_runner.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dtrai2) [reply tests/unit/util/test_rule_dry_runner.py:179]: ah okay, yeah I agree we can fix this in a separate pull-request then. :+1:

File: tests/unit/exceptions/processor/test_processing_warning.py
[the remark this answers is not in the document]
Comment (dtrai2) [reply tests/unit/exceptions/processor/test_processing_warning.py:1]: okay, fair enough!

File: logprep/abc/input.py [hunk 16]
```diff
@@ -23,26 +23,42 @@
...
 
 class FatalInputError(InputError):
     """Must not be catched."""
 
+    def __init__(self, input_connector: "Input", message: str) -> None:
+        input_connector.metrics.number_of_errors += 1
+        super().__init__(input_connector, message)
+
 
 class WarningInputError(InputError):
```
[the remark this answers is not in the document]
Comment (dtrai2) [reply logprep/abc/input.py:47]: haha that's true for all variables. So I guess you don't mind if I start naming mine "var1", "var2" and so on :wink: But yeah I agree that it is a minor remark.

**Review by ppcad — APPROVED** [review]
Thanks, it does not break on my site.

**Pull request merged**